### ISSUE TITLE
Can't use VoiceConnection when the bot is moved to another vc by a User

### ISSUE THREAD

**Issue by Mexican-Man** [bwmarrin/discordgo#849]
Ran into a situation where when playing a sound, if the bot was moved to another voice channel, it seemingly gets stuck in limbo. Running `Disconnect()` just crashes. `s.VoiceConnections[e.GuildID]` seems to just flat out return `nil` every time (for me at least).

I was able to sorta get around this by storing the `VoiceConnection*` as a global, then setting a custom error handler for dgvoice:
```go
dgvoice.OnError = func(str string, err error) {
	vc.Disconnect()
}
```
then setting up an event handler for `voiceStateUpdate`:
```go
if e.UserID == "bot-id" {
	if e.BeforeUpdate != nil {
		vc, _ = s.ChannelVoiceJoin(e.GuildID, e.ChannelID, false, false)
		vc.Disconnect()
	}
	return
}
```
This also requires #833 in order to work, or some manual work keeping track of the bot's previous voice channel.

Basically, if it detects the bot being moved (`e.BeforeUpdate != nil`) then it reconnects to the same channel manually, bringing it out of "limbo". Then it just disconnects.

Obviously this isn't the end of the world, but it's a pretty rough fix. Does anyone have any better approaches? My ideas are:
- Add some "fromAnotherChannel" property to `VoiceStateUpdate`, then it's at least a bit easier for users to define their wanted behavior
- Add a separate "voiceChannelMoved` event to make it even a little bit _more_ easy
- Ideally, somehow have `VoiceConnection*` update itself when moved

Thoughts?

**Comment by nonnonstop** [comment]
I'm having same issue.

When I moved the bot to another VC, I got the following log.
```
...
2021/03/07 18:56:55 [DG2] voice.go:354:wsListen() received 4014 manual disconnection
2021/03/07 18:56:55 [DG2] voice.go:167:Close() called
2021/03/07 18:56:55 [DG2] voice.go:176:Close() closing v.close
2021/03/07 18:56:55 [DG2] voice.go:182:Close() closing udp
...
```

This log is output from voice.go
https://github.com/bwmarrin/discordgo/blob/37088aefec2241139e59b9b804f193b539be25d6/voice.go#L351-L368

I found the same issue at Rapptz/discord.py#5904.
Maybe the connection check is still needed when the client receives 4014 error.

**Comment by qaisjp** [comment]
> storing the `VoiceConnection*` as a global, then setting a custom error handler for dgvoice:
> 
> ```go
> dgvoice.OnError = func(str string, err error) {
> 	vc.Disconnect()
> }
> ```

Btw, as a general Go tip, you need not store it as a global.

If you have something like this:

```go
type MyBot struct {
	d *discordgo.Session
	vc *discordgo.VoiceConnection
}

func (b *MyBot) onError(str string, err error) {
	b.vc.Disconnect()
}
```

You can assign a method "bound" to this struct like this:

```go
func main() {
	bot := &MyBot{}
	// assign some things inside bot
	dgvoice.OnError = bot.onError
}
```

**Comment by Crypto90** [comment]
Any news about this?
